File: 1122-relative-sort-array/custom-sort/java/solution.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        // Map for rank lookups
        Map<Integer, Integer> order = new HashMap<>();  
        for (int idx = 0; idx < arr2.length; idx++) 
            order.put(arr2[idx], idx);
        
        // Explicit Boxing (Required for custom Comparator support)
        Integer[] boxed = new Integer[arr1.length]; 
        for (int idx = 0; idx < arr1.length; idx++) 
            boxed[idx] = arr1[idx];

        // Custom Comparator
        Comparator<Integer> customComparator = new Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                int rankA = order.getOrDefault(a, Integer.MAX_VALUE);
                int rankB = order.getOrDefault(b, Integer.MAX_VALUE);
    
                return (rankA != rankB) ? rankA - rankB : a - b;
            }
        };

        // Sorting 
        Arrays.sort(boxed, customComparator);

        // Explicit Unboxing
        for (int idx = 0; idx < arr1.length; idx++)
            arr1[idx] = boxed[idx];
        
        return arr1;
    }
}

File: 1122-relative-sort-array/custom-sort/cpp/solution.cpp
struct CustomOrder {
    const unordered_map<int, int>& orderMap;
    const int INF = numeric_limits<int>::max();

    CustomOrder(const unordered_map<int, int>& m) : orderMap(m) {}

    bool operator()(int a, int b) const {
        auto itA = orderMap.find(a);
        auto itB = orderMap.find(b);

        // If in arr2, use its index, otherwise, treat as INF
        int rankA = (itA != orderMap.end()) ? itA->second : INF;
        int rankB = (itB != orderMap.end()) ? itB->second : INF;

        // Primary sort by arr2 index, secondary sort by value (asc).
        return (rankA != rankB) ? rankA < rankB : a < b; 
    }
};

class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        unordered_map<int, int> order;
        for (int idx = 0; idx < arr2.size(); idx++) 
            order[arr2[idx]] = idx;

        sort(arr1.begin(), arr1.end(), CustomOrder(order));
        return arr1;     
    }
};
